File: frontend/src/utils/distanceUnits.ts
import type { DistanceUnit } from "../types/exercise";

const KM_PER_MILE = 1.609344;

export function convertDistance(value: number, from: DistanceUnit, to: DistanceUnit): number {
  if (from === to) return value;
  return from === "mi" ? value * KM_PER_MILE : value / KM_PER_MILE;
}

export function formatDistance(kilometers: number, unit: DistanceUnit): string {
  return `${convertDistance(kilometers, "km", unit).toFixed(2)} ${unit}`;
}

File: backend/app/services/distance_units.py
from decimal import Decimal, ROUND_HALF_UP

from app.models.exercise import DistanceUnit

KM_PER_MILE = Decimal("1.609344")
DISTANCE_STORAGE_PRECISION = Decimal("0.001")


def to_kilometers(distance: Decimal, unit: DistanceUnit) -> Decimal:
    kilometers = distance if unit is DistanceUnit.KILOMETERS else distance * KM_PER_MILE
    return kilometers.quantize(DISTANCE_STORAGE_PRECISION, rounding=ROUND_HALF_UP)
